### PR TITLE
Added Advanced Controller Mode with multiple Menu options

### PR DESCRIPTION
**Pull request by teinturman**
The advanced Controller can be enabled by setting the CONTROLLER_ADVANCED=True in the myconfig.py file. 

If config file is not updated, then the code run will be identical to the previous version of controller.py.

The advanced controller provides multiple modes : select L1, then R1 to change the mode. 
SYSTEM MODE : menu to display system information
CAMERA MODE : menu to calibrate a pan / tilt turret
CALIBRATE : menu to calibrate easily steering and throttle settings

The HOME2 button will display each menu options.

The Training mode allows to change Behaviours and localization via the pad : 
TRAINING (R1) : advanced training options .

Features : 
- wiiu controls updated to add the Home button
- Advanced controls are identical for all controllers, (PS3,PS4, WIIU etc...)
- The display of information on screen is also sent to an input variable that could be used by a part controlling an oled display in the car. 
- In the SYSTEM Mode , a value for AUG_IMAGE_DRAW LEVEL has been planned ::  a part using this value could draw additional information on the captured images. 
This could be used to debug  / calibrate the car, or used in conjonction with making a video.
0 : No augmented drawing
1 : CROP visualisation
2 : Steering / Throttle visualisation
3 : Pilot_Behaviour and Pilot_localization
4 : User_Behaviour and User_Localization
5 : all drawings between 1 and 4
6 -9:  reserved for future usage
10 : all drawings between 6 and 9

- The HOME2 button displays the help for each mode.
- Updated documentation on controller is uploaded, with images showing each menu options.

### CODE REVIEW

File: docs/parts/controllers.md [hunk 1]
```diff
@@ -223,6 +223,52 @@ quit
 
 Now that your controller is trusted, it should automatically connect with your Raspberry Pi when they are both turned on.  If your controller fails to connect, run the bluetoothctl steps again to reconnect.
 
+## Advanced Joystick Mode
+
+The Advanced Joystick mode uses button combinations to allow more control on the car.
+In order to use the Advanced Joystick mode, you must update the manage.py file and  replace the line get_controller_js() with get_controller_js(Vobject=V)
```
Comment (nvtkaszpir) [inline docs/parts/controllers.md]: Please wrap code parts in backtics, such as: `some code`

File: docs/parts/controllers.md [hunk 2]
```diff
@@ -223,6 +223,52 @@ quit
...
+You will also need to modify the config.py file to add the line "JOYSTICK_ADVANCED = True 
+Setting the JOYSTICK_ADVANCED = False will use the original Joystick mappings, without advanced feature.
+When activated , this mode will redefine the mappings of the joystick as per the below description :
+
+When using Advanced mode, pressing combinations of L1 and R1 button will switch the controller to one of the 5 different modes :  Default (), TRAIN (R1) , CAMERA(L1), CALIBRATE(R1+L1) and SYSTEM(L1+R1).
+Please note that not all features are working yet;the modes anticipate the creation of custom parts handling the actions.
+
+
+![Default Mode](./controller_default.PNG)
```
Comment (nvtkaszpir) [inline docs/parts/controllers.md]: nit picking: lowercase file extensions

File: donkeycar/parts/controller.py [hunk 3]
```diff
@@ -659,6 +660,13 @@ def __init__(self, poll_delay=0.0,
         self.axis_trigger_map = {}
         self.init_trigger_maps()
 
+        self.shiftKeyHistory=[]
```
Comment (nvtkaszpir) [inline donkeycar/parts/controller.py]: Please follow code formatting as it is presented in lines above (spaces around =)

File: docs/parts/controllers.md (hunk 1 above)
Comment (teinturman) [reply docs/parts/controllers.md]: Hi, thanks for your feedback, i believe i have resolved the point now with my new controller.py  ( 21/10/2019). 
Regards,

File: docs/parts/controllers.md (hunk 2 above)
Comment (teinturman) [reply docs/parts/controllers.md]: Hi, i believe i have resolved.

File: donkeycar/parts/controller.py (hunk 3 above)
Comment (teinturman) [reply donkeycar/parts/controller.py]: Hi, i have updated all occurences  and added spaces before and after '='. 
thx.

File: docs/parts/controllers.md [hunk 4]
```diff
@@ -223,6 +223,57 @@ quit
 
 Now that your controller is trusted, it should automatically connect with your Raspberry Pi when they are both turned on.  If your controller fails to connect, run the bluetoothctl steps again to reconnect.
 
+
+## Advanced Joystick Mode
+
+The Advanced Joystick mode uses button combinations to allow more control on the car.
+In order to use the Advanced Joystick mode, you must update the manage.py file and  replace the line `get_controller_js()` with `get_controller_js(Vobject=V)`
+You will also need to modify the config.py file to add the line `JOYSTICK_ADVANCED = True `
+Setting the JOYSTICK_ADVANCED = False will use the original Joystick mappings, without advanced feature.
```
Comment (DocGarbanzo) [inline docs/parts/controllers.md]: back ticks missing

File: donkeycar/parts/controller.py [hunk 5]
```diff
@@ -791,6 +800,229 @@ def update(self):
 
             time.sleep(self.poll_delay)
 
+    def update_adv(self):
+        '''
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: pep-8 wants double triple quotes for doc string

File: donkeycar/parts/controller.py [hunk 6]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+        poll a joystick for input events
+        '''
+        self.shiftkey=""
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: missing blanks

File: donkeycar/parts/controller.py [hunk 7]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+        while self.running and self.js is None and not self.init_js():
+            time.sleep(3)
+
+        while self.running:
+            button, button_state, axis, axis_val = self.js.poll()
+            if axis is not None :
+                if axis in self.key_translate_map:      
+                   axis=self.key_translate_map[axis]  # This would not be needed if we standardized the axis names... 
+               
+                shifted_axis=self.shiftkey+axis
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: missing blanks

File: donkeycar/parts/controller.py [hunk 8]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+                if  shifted_axis in self.axis_trigger_map:
+                   if type(self.axis_trigger_map[shifted_axis]) is str :
+                       funcstr=self.axis_trigger_map[shifted_axis]
+                       funcstr2=funcstr.replace("%val%",str(axis_val))
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: missing blanks in these 2 lines, too

File: donkeycar/parts/controller.py [hunk 9]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+                       exec(funcstr2)
+                   else:
+                      self.axis_trigger_map[shifted_axis](axis_val)
+                else:
+                   if axis[:1]=="|":
+                       k,dp1,dp2= axis.split("|")
+                       button,button_state= self.buttonize(axis_val,dp1,dp2,dp1+"_"+dp2)
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: .. and here, too.

File: donkeycar/parts/controller.py [hunk 10]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+                       if type(self.button_down_trigger_map[button]) is str :
+                           funcstr = self.button_down_trigger_map[button]
+                           exec(funcstr)
+                       else:
+                           self.button_down_trigger_map[button]
+ 
+                     
+
+            if button and button_state == 0:
+                '''
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: In-function comment should use `#`

File: donkeycar/parts/controller.py [hunk 11]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+            nm = 0
+        else : 
+            nm = i+1
+        self.menu_in(self.menulist[nm],self.menudescr[nm])
+
+    def menu_start(self,btn):
+        self.MenuKey=btn
+        self.shift_clear
+        self.menu_in(self.menulist[0],self.menudescr[0])
+
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: Too many empty lines

File: donkeycar/parts/controller.py [hunk 12]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+        
+    def menu_register(self,MenuString,button_down_trigger_map,button_up_trigger_map,axis_trigger_map,descr = ""):
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: missing blanks

File: donkeycar/parts/controller.py [hunk 13]
```diff
@@ -791,6 +800,229 @@ def update(self):
...
+            self.button_down_trigger_map[t] = button_down_trigger_map[t]
+        for t in axis_trigger_map:
+            self.axis_trigger_map[t] = axis_trigger_map[t] 
+
+
+    def shift_in(self,btn,Mode,descr = ""):
+        self.shiftKeyHistory.append(btn)
+        self.shiftModeHistory[btn] = self.shiftkey
+        self.shiftkey = Mode+"+"
+        if descr != "" :print(descr)
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py]: missing line break

**Comment by DocGarbanzo** [comment]
@teinturman - you might want to load your code into PyCharm or any other IDE which will show you all pep-8 compliance breaches, missing blanks, too many blanks, missing line breaks, doc string comments, etc.

**Comment by teinturman** [comment]
This change is now outdated. I close this PR

**Pull request closed without merging**